pysys/process/monitor.py

```diff
@@ -0,0 +1,454 @@
+		# get an immediate error if it fails
+		self.thread = self.owner.startBackgroundThread(str(self), self.__backgroundThread)
+		return self
+	
+	def __str__(self): return 'ProcessMonitor<%s>'%('%s pid=%d'%(self.process,self.pid) if self.process else self.pid)
+	
+	def _preprocessData(self, data):
+		""" Called in the background thread with the data dictionary from 
+		each poll of the process, to allow OS-agnostic pre-processing and 
+		addition of derived data keys such as the date and time. 
+		
+		@param data: The dictionary of process monitoring data. This method 
+		may add or modify the contents of this dictionary. 
+		
+		"""
+		
+		datetime = time.gmtime(time.time())
+		data[ProcessMonitorKey.DATE_TIME] = time.strftime("%Y-%m-%d %H:%M:%S", datetime)
+		data[ProcessMonitorKey.DATE_TIME_LEGACY] = time.strftime(
+			"%d/%m/%y %H:%M:%S" if IS_WINDOWS else "%m/%d/%y %H:%M:%S", datetime)
+		
+		if data.get(ProcessMonitorKey.CPU_CORE_UTILIZATION,None) is not None:
+
+			data[ProcessMonitorKey.CPU_TOTAL_UTILIZATION] = int(data[ProcessMonitorKey.CPU_CORE_UTILIZATION] / self._cpuCount)
+			
+			if self.__numProcessors > 1:
+				# undocumented, for compatibility only
+				data[ProcessMonitorKey.CPU_CORE_UTILIZATION] = data[ProcessMonitorKey.CPU_CORE_UTILIZATION] / self.__numProcessors
+	
+	def __backgroundThread(self, log, stopping):
+		self._stopping = stopping # for use by other process monitor methods
+		sample = 1
+		try:
+			while not stopping.is_set():
+				d = self._getData(sample)
+				assert d, 'No data returned'
+				
+				d[ProcessMonitorKey.SAMPLE] = sample
+				self._preprocessData(d)
+				
+				for h in self.handlers:
+					h.handleData(d)
+				sample += 1
+				stopping.wait(self.interval)
+		except Exception as ex:
+			if self.process and not self.process.running():
+				log.debug('Ignoring process monitor error as the monitored process %s has already terminated: %s', self.process, ex)
+			else:
+				raise
+		finally:
+			log.debug('Calling cleanup on process monitor handler(s)')
+			try:
+				for l in self.handlers:
+					if hasattr(l, 'cleanup'): l.cleanup()
+			finally:
+				self._cleanup()
+
+	def running(self):
+		"""Return the running status of the process monitor.
+		
+		@return: True if the process monitor background thread is still running. 
+		@rtype: bool
+		"""
+		return self.thread.is_alive()
+
+	
+	def stop(self):
+		"""Request the process monitor thread to terminate.
+		
+		Does not wait for the termination to complete. 
+		
+		"""
+		self.thread.stop()
+
+	# for implementation by subclasses
+
+	def _getData(self, sample):
+		"""Implement gathering of latest monitoring data. 
+		
+		Called on the background monitoring thread regularly.
+		
+		@param sample: An integer starting at 1 and incrementing each time 
+		this method is called. 
+		
+		@return: A dictionary of (typically numeric) values, keyed by 
+		L{ProcessMonitorKey}.
+		@rtype: dict
+		"""
+		raise NotImplementedError('_getData is not implemented yet')
```

Comment:
Maybe a better message would be that "_getData must be implemented in a subclass" or something similar?  The "yet" implies there might one day be an implementation here which I think is unlikely.

Reply:
yes fixed